Fix NoiseTexture2D.get_image() returning null on first call

ytianle opened the pull request:
Ensure the internal image is generated before the first get_image() call. This prevents get_image() from returning null in _ready() unless manually awaited.

Fixes godotengine/godot#105261

ytianle committed:
Fix NoiseTexture2D.get_image() returning null on first call

Ensure the internal image is generated before the first get_image() call. This prevents get_image() from returning null in _ready() unless manually awaited.

yuantianle commented:
Demoing:
![image](https://github.com/user-attachments/assets/3b1bd02b-fd4c-41a6-98f3-ab59707af7ce)

beicause reviewed (commented):
Seems not right. `_update_texture()` just starts a thread to update image, so the returned result may still be null.

kleonc commented:
Also `const_cast`-ing like currently done in this PR can be buggy, see #93759/#94744.

ytianle committed:
Add _generate_image_now private func

`NoiseTexture2D.get_image()` returns null when called in `_ready()` because the texture image hasn't been generated yet. This change makes `get_image()` synchronously generate the image (via `_generate_image_now()`) on first access, ensuring it's available when needed immediately.

yuantianle commented:
@kleonc since the image creation is generated under the read-only func, I could not find another way to set texture image without using `const_cast`

kleonc commented:
> @kleonc I have uploaded my update with the newest commit. Since there is no image initialized when we call `get_image()`, we will need to synchronously generate the image. However, if we want to put the `write` (_set_texture_image()) under the `read-only` function `get_image()`, using `const_cast` is the only way shown in my mind.
> 
> Please let me know if you have a better way! Thanks!

You can do what was done in #94744 (3e0d3c433bde1f2e0b18c3f05e43c3bf3c0804bc), which instead made the updating/generating fuctions be `const` (and thus callable without `const_cast` from within `get_image`).

yuantianle commented:
> > @kleonc I have uploaded my update with the newest commit. Since there is no image initialized when we call `get_image()`, we will need to synchronously generate the image. However, if we want to put the `write` (_set_texture_image()) under the `read-only` function `get_image()`, using `const_cast` is the only way shown in my mind.
> > Please let me know if you have a better way! Thanks!
> 
> You can do what was done in #94744 ([3e0d3c4](https://github.com/godotengine/godot/commit/3e0d3c433bde1f2e0b18c3f05e43c3bf3c0804bc)), which instead made the updating/generating fuctions be `const` (and thus callable without `const_cast` from within `get_image`).

I have already tried this way, however, it is a method inherited from Texture2D. If we do not follow the signature, when the script is calling `get_image()`, our method with the new signature won't be triggered. I had already tested that,

kleonc commented:
> I have already tried this way, however, it is a method inherited from Texture2D. If we do not follow the signature, when the script is calling `get_image()`, our method with the new signature won't be triggered (it somehow calls the parent). I had already tested that. Please let me know if I understood sth wrong. Thanks.

`get_image` is inherited and its signature should be not changed indeed. :+1:

What I meant is that changing e.g. these functions to `const` should be fine:
https://github.com/godotengine/godot/blob/7b9c5122faafb6b4a655e31ae27bcffeb5cf254f/modules/noise/noise_texture_2d.h#L70-L71

Just like e.g. for GradientTexture1D these were changed to `const` (in #94744):
https://github.com/godotengine/godot/blob/7b9c5122faafb6b4a655e31ae27bcffeb5cf254f/scene/resources/gradient_texture.h#L46
https://github.com/godotengine/godot/blob/7b9c5122faafb6b4a655e31ae27bcffeb5cf254f/scene/resources/gradient_texture.h#L66
and thus it can be called from `get_image` without `const_cast`:
https://github.com/godotengine/godot/blob/7b9c5122faafb6b4a655e31ae27bcffeb5cf254f/scene/resources/gradient_texture.cpp#L174-L180

yuantianle commented:
If we follow the way you are mentioning, how can we solve the issue with this `_set_texture_image` function? https://github.com/godotengine/godot/blob/7b9c5122faafb6b4a655e31ae27bcffeb5cf254f/scene/resources/gradient_texture.cpp#L61
This is the difference between `noise_texture_2d.cpp` and `gradient_texture.cpp`

beicause commented:
It's true that this is different from GradientTexture. 
I left a comment in https://github.com/godotengine/godot/issues/105261#issuecomment-2797655130

yuantianle commented:
No worries at all! Thanks for the suggestions, though!
